### DISCUSSION TITLE
Issue with changing focus to specified Input field with on_key enter while Button is focused

### DISCUSSION THREAD

**Discussion by apeaveysf** [Textualize/textual#1197] [Q&A] [answered]
I have added functionality to a small app to move from field to field with the enter key instead of tab, and when a button is the next field, execute the action for that button and move to the next appropriate Input field.

I've minimized the code to essentials in attached to show the behavior in Reset button of this app moving to the #field_client_id when clicked with the mouse as expected, but when using the enter key when the Reset button is focused, it moves to #field_client_name instead.

[demo_printer.zip](https://github.com/Textualize/textual/files/10023992/demo_printer.zip)

Discussion in Discord #help-wanted channel just prior to this posting.

For full code (does not have json file code and custom module stripped and likely will not run without them): [https://github.com/apeaveysf/demo_label_printer](url)

**Comment by davep** [comment]
Thanks! As mentioned on Discord, I think there's a tidy workaround to be had here, and perhaps some changes to Textual in the longer-term that will make this easier and more obvious.

End of office hours for me right now but I hope to return to this later this evening or in the morning. :-)

**Comment by davep** [comment, marked as answer]
Looking at this a bit more now, the first thing I'd suggest is that, rather than hooking into the key event to turn `Enter` into the "move to next field" method, it might make a bit more sense to use [the input submitted event](https://textual.textualize.io/widgets/input/#submitted).  Possibly not the most elegant way, but a quick first play here suggests something like this perhaps:

```python
    def on_input_submitted(self, event: Input.Submitted ) -> None:
        next_input = {
            "field_label_printer": "client_id",
            "field_client_id": "client_name",
            "field_client_name": "alias",
            "field_alias": "tests",
            "field_tests": "date",
            "field_date": "quantity",
            "field_quantity": "label_printer"
        }[ event.input.id ]
        self.set_focus( self.query_one( f"#field_{next_input}", Input ) )
```

As for the specific issue, if focus going to the second field when `Enter` is used on the Reset button: my initial guess here is that the `Enter` event that arises from using the keyboard to "click" Reset is bleeding over into the newly-focused input field, which then in turn processes that keyboard event. You can see that your `action_clear_fields` method does what you'd expect if you use the mouse to click on the reset button.

This is better illustrated if I add some debugging to the `on_input_submitted` method and the `action_clear_fields` method. If I do this and use the mouse to click on the reset button, I just get this:

<img width="179" alt="Screenshot 2022-11-16 at 21 18 21" src="https://user-images.githubusercontent.com/28237/202296711-7d8490ec-8985-4153-9570-dcecf4655452.png">

On the other hand, if I use the enter key on the reset button, it really does look like focus moves to the first field, as per `action_clear_fields`, and then the enter key is processed:

<img width="756" alt="Screenshot 2022-11-16 at 21 20 34" src="https://user-images.githubusercontent.com/28237/202296920-88dbf8bc-5a15-4b85-8d04-69954bc1bd60.png">

While something like #1196 would greatly simplify what I did in the (kinda klunky) implementation of `on_input_submitted` above, I doubt it would help really solve the issue you show here. This looks like a bit of a bug to me, possibly.

I'll narrow this down to the smallest test example possible and raise an issue for us to look at.

**Reply by davep** [reply]
I've now raised #1199 for the core problem here.

**Reply by apeaveysf** [reply]
Thanks, I did not know about on_input_submitted, and agree it is more appropriate.  Is there an easy list of all on_ options; or is basically on_widgetname_<various messages from the reference?

After learning the default screen exists and figuring out it's id is #_default, I took great inspiration from your suggestion/snippet, I simplified it to:

```
def on_input_submitted(self, event: Input.Submitted) -> None:
    """Event handler for Input field submissions."""
    input_control = {"next": ["field_label_printer", "field_client_name", "field_alias", "field_tests", "field_date"],
                "field_client_id": self.action_load_client,
                "field_quantity": self.action_print_label}
                
    if self.focused.id in input_control.get("next"):
        self.query_one("#_default").focus_next()
    else:
        input_control.get(self.focused.id)()
```

Threw in method calls for Inputs that should perform an action.

Looking at your... dev console? log? (i've not used the dev console yet), I agree, that looks like a bug.  I will come back after a release or two to check if it is resolved.

**Reply by davep** [reply]
> Is there an easy list of all on_ options; or is basically on_widgetname_<various messages from the reference?

You should find that each widget in the reference section of the Textual manual, where it makes sense, has a "Messages" section listing them. So see the [Messages section of the reference for `Input` for example](https://textual.textualize.io/widgets/input/#messages).

> Looking at your... dev console? log? (i've not used the dev console yet), I agree, that looks like a bug. I will come back after a release or two to check if it is resolved.

If/when it gets resolved I'll do my best to try and remember to update this discussion to give you a heads-up; and of course you can subscribe to #1199 to keep track of any work on it.
